File: src/hangman.py
import random
import sys


def main():
    print("=== HANGMAN ===")

    if len(sys.argv) < 2:
        print("Please provide a file path.", file=sys.stderr)
        sys.exit(1)
    filename = sys.argv[1]

    try:
        file = open(filename, "r").read()
    except Exception:
        print("Failed to read file.", file=sys.stderr)
        sys.exit(2)

    words = []
    for line in file.split("\n"):
        line = line.strip()
        if len(line) > 0:
            words.append(line)

    print("\n\n\n\n\n")

    while True:
        index = random.randint(0, (len(words) - 1))
        word = words[index]

        correct = set()
        incorrect = set()

        while True:
            for i in range(0, 5):
                print("\033[A\033[K", end="")

            visible = ""
            is_win = True
            for ch in word:
                if ch in correct:
                    visible += ch
                else:
                    visible += "_"
                    is_win = False

            if is_win:
                print("---------")
                print("You win! :)")
                print(f"The word was: '{word}'")
                print("---------")
                input()
                break
            if len(incorrect) >= 6:
                print("---------")
                print("You lose! :(")
                print(f"The word was: '{word}'")
                print("---------")
                input()
                break

            print(visible)
            print("Chances:", 6 - len(incorrect))
            print("Correct:", ", ".join(correct))
            print("Incorrect:", ", ".join(incorrect))

            line = input("Guess: ")
            if len(line) < 1:
                continue
            guess = line[0]

            if guess in word:
                correct.add(guess)
            else:
                incorrect.add(guess)


if __name__ == "__main__":
    main()

File: src/hangman.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"math/rand"
	"os"
	"strings"
)

func main() {
	fmt.Println("=== HANGMAN ===")

	stderr := log.New(os.Stderr, "", 0)

	if len(os.Args) < 2 {
		stderr.Println("Please provide a file path.")
		os.Exit(1)
	}
	filename := os.Args[1]

	file, err := os.ReadFile(filename)
	if err != nil {
		stderr.Println("Failed to read file.")
		os.Exit(2)
	}
	words := strings.Split(string(file), "\n")

	scanner := bufio.NewScanner(os.Stdin)

	fmt.Print("\n\n\n\n\n\n")

	for {
		index := rand.Intn(len(words))
		word := strings.TrimSpace(words[index])

		correct := make(map[rune]bool)
		incorrect := make(map[rune]bool)

		for {
			for i := 0; i < 5; i++ {
				fmt.Print("\033[A\033[K")
			}

			visible := ""
			is_win := true
			for _, ch := range word {
				if correct[ch] {
					visible += string(ch)
				} else {
					visible += "_"
					is_win = false
				}
			}

			if is_win {
				fmt.Println("---------")
				fmt.Println("You win! :)")
				fmt.Printf("The word was: '%s'\n", word)
				fmt.Println("---------")
				scanner.Scan()
				break
			}
			if len(incorrect) >= 6 {
				fmt.Println("---------")
				fmt.Println("You lose! :(")
				fmt.Printf("The word was: '%s'\n", word)
				fmt.Println("---------")
				scanner.Scan()
				break
			}

			fmt.Println(visible)
			fmt.Printf("Chances: %d\n", 6-len(incorrect))
			fmt.Print("Correct: ")
			printSet(correct)
			fmt.Print("Incorrect: ")
			printSet(incorrect)
			fmt.Print("Guess: ")

			var line string
			fmt.Scanln(&line)
			if len(line) < 1 {
				continue
			}
			guess := []rune(line)[0]

			if strings.ContainsRune(word, guess) {
				correct[guess] = true
			} else {
				incorrect[guess] = true
			}
		}
	}
}

func printSet(set map[rune]bool) {
	i := 0
	for ch, _ := range set {
		if i > 0 {
			fmt.Print(", ")
		}
		fmt.Printf("%c", ch)
		i++
	}
	fmt.Println()
}
